src/core/include/mp-units/framework/value_cast.h

```diff
@@ -152,4 +178,106 @@ template<Unit auto ToU, Representation ToRep, typename QP>
+           (detail::same_absolute_point_origins(ToQP::point_origin, std::remove_reference_t<QP>::point_origin)) &&
+           std::constructible_from<typename ToQP::rep, typename std::remove_reference_t<QP>::rep>
+[[nodiscard]] constexpr QuantityPoint auto value_cast(QP&& qp)
+{
+  using qp_type = std::remove_reference_t<QP>;
+  if constexpr (is_same_v<std::remove_const_t<decltype(ToQP::point_origin)>,
+                          std::remove_const_t<decltype(qp_type::point_origin)>>) {
+    return quantity_point{
+      value_cast<typename ToQP::quantity_type>(std::forward<QP>(qp).quantity_from(qp_type::point_origin)),
+      qp_type::point_origin};
+  } else {
+    // it's unclear how hard we should try to avoid truncation here. For now, the only corner case we cater for,
+    // is when the range of the quantity type of at most one of QP or ToQP doesn't cover the offset between the
+    // point origins. In that case, we need to be careful to ensure we use the quantity type with the larger range
+    // of the two to perform the point_origin conversion.
+    // Numerically, we'll potentially need to do three things:
+    //  (a) cast the representation type
+    //  (b) scale the numerical value
+    //  (c) add/subtract the origin difference
+    // In the following, we carefully select the order of these three operations: each of (a) and (b) is scheduled
+    // either before or after (c), such that (c) acts on the largest range possible among all combination of source
+    // and target unit and represenation.
+    constexpr Magnitude auto c_mag = get_canonical_unit(qp_type::unit).mag / get_canonical_unit(ToQP::unit).mag;
+    constexpr Magnitude auto num = detail::numerator(c_mag);
+    constexpr Magnitude auto den = detail::denominator(c_mag);
+    constexpr Magnitude auto irr = c_mag * (den / num);
+    using c_rep_type = detail::maybe_common_type<typename ToQP::rep, typename qp_type::rep>;
+    using c_mag_type = detail::common_magnitude_type<c_mag>;
+    using multiplier_type = conditional<
+      treat_as_floating_point<c_rep_type>,
+      // ensure that the multiplier is also floating-point
+      conditional<std::is_arithmetic_v<value_type_t<c_rep_type>>,
+                  // reuse user's type if possible
+                  std::common_type_t<c_mag_type, value_type_t<c_rep_type>>, std::common_type_t<c_mag_type, double>>,
+      c_mag_type>;
+    constexpr auto val = [](Magnitude auto m) { return get_value<multiplier_type>(m); };
+    if constexpr (val(num) * val(irr) > val(den)) {
+      // original unit had a larger unit magnitude; if we first convert to the common representation but retain the
+      // unit, we obtain the largest possible range while not causing truncation of fractional values. This is optimal
+      // for the offset computation.
+      return value_cast<typename ToQP::quantity_type>(
+        value_cast<c_rep_type>(std::forward<QP>(qp)).point_for(ToQP::point_origin));
```

Comment:
Can we reuse `sude_cast` here? I wouldn't like to duplicate the implementation. Maybe `sudo_cast` could benefit from new logic as well?

Reply:
`sudo_cast` is being used (indirectly), through the two `value_cast`. The apparent duplication is just in order to determine the conversion factor to be applied to the numeric value. The runtime code in the body here does not apply that conversion factor at all; but it needs to know that factor is eventually going to make number larger or smaller. That is needed in turn to decide when to do that conversion: if the number gets larger, we want to postpone until after the offset adjustment, and if it gets smaller, we want to do it before the offset adjustment.

We could probably create a second overload of `sudo_cast` which operates on `quantity_point` (basically the content of this `value_cast`, minus the constraint on matching quantity_specs, and ensuring that the quantity_spec is being casted correctly). Then, we could extract that calculation of the rescaling factor into a helper function and use it from both overloads.

Reply:
I would prefer that. I've already learned that `sudo_cast` is really fragile, and any changes should be verified against the codegen in Compiler Explorer. I would not like to have nearly copy pasted implementation in a separate file because it would be hard to maintain and keep consistent.

Reply:
@burnpanck, I hope that you successfully delivered the upgraded project on time. Do you still work on this PR?

Reply:
The refactor is done, but we had to postpone a few features, which we are still delivering as we speak. I do intend to finish this one tough, as our code is already heavily reliant on these `value_cast` variants.

Reply:
I plan to release mp-units 2.2 in the following days. Please let me know if you will have time to finish it soon.

Reply:
On it. I'm currently refactoring the implementation of the new `value_cast<QP>` into a new overload of `sudo_cast`, sharing the multiplier calculation between the two `sudo_cast` overloads as you prefer. While doing so, I noticed a few details on the existing `sudo_cast<Q>` that appear disputable to me. I may open a separate issue to review that with you, but for now, I'll just follow by analogy.

Reply:
As far as the code sharing between the two `sudo_cast` overloads, regarding

> I would prefer that. I've already learned that sudo_cast is really fragile, and any changes should be verified against the codegen in Compiler Explorer. I would not like to have nearly copy pasted implementation in a separate file because it would be hard to maintain and keep consistent.

The shared code is purely compile-time; all of the runtime behaviour of `sudo_cast<Q>` was already re-used within my previous `value_cast<QP>` implementation through the call to `value_cast<Q>`. Now, the new `sudo_cast<QP>` will share the compile-time computation through a separate helper, and re-use the runtime code through a call to `sudo_cast<Q>`. `sudo_cast<Q>` will have most of it's compile-time guts ripped out into that shared helper, but it's runtime-path will remain untouched.

Reply:
I didn't mean the final result might be wrong. It never was and we unit test for it as well.

What I meant here is that the same conversion may compile into more assembly instructions than a similar operation on a fundamental type. I do not know how to unit test this, and I always discover an issue only while preparing some slides to my talks that prove that there is not overhead over using `double`. Unfortunately, I often find this to not be true anymore in this case and have to fix the `sudo` cast.

This is why I would like you to check the assembly before merging. Alternatively, we can merge it, and then verify in the Compiler Explorer and fix things if the codegen will be worse.

Reply:
Sure, but I did not touch the original `sudo_cast` implementation of that multiplication, and the `sudo_cast<QP>` implementation does not do any unit conversion on it's own. It fully delegates to that to the existing `sudo_cast<Q>`.

Reply:
Sure! This is why I asked to have one place where those things are implemented. Initially, you copy-pasted it to another file, which would make it much harder to maintain and ensure synergy and correctness.